File: ITMO.Python/PythonProject/Demo02/benchmark/java/src/Engine/DoubleMathematics.java
package Engine;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class DoubleMathematics implements IMathematics<Double> {
    @Override
    public Double[][] transposeMatrix(Double[][] matrix) {
        int rows = matrix.length;
        int columns = matrix[0].length;

        Double[][] transposedMatrix = new Double[columns][rows];
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < columns; col++) {
                transposedMatrix[col][row] = matrix[row][col];
            }
        }

        return transposedMatrix;
    }

    @Override
    public Double[][] multiplyMatrices(Double[][] matrixA, Double[][] matrixB) {
        Double[][] result = new Double[matrixA.length][matrixB[0].length];
        for (int row = 0; row < matrixA.length; row++) {
            for (int col = 0; col < matrixB[0].length; col++) {
                for (int idx = 0; idx < matrixB.length; idx++) {
                    result[row][col] = result[row][col] == null ?
                            matrixA[row][idx] * matrixB[idx][col]
                            : result[row][col] + matrixA[row][idx] * matrixB[idx][col];
                }
            }
        }
        return result;
    }

    @Override
    public Double[][] multiplyMatricesParallel(Double[][] matrixA, Double[][] matrixB, int threadCount) {
        Double[][] result = new Double[matrixA.length][matrixB[0].length];

        ExecutorService executor = Executors.newFixedThreadPool(threadCount);
        CountDownLatch latch = new CountDownLatch(matrixA.length);
        for (int idx = 0; idx < matrixA.length; idx++) {
            int row = idx;
            executor.submit(new Runnable() {
                public void run() {
                    synchronized (result){
                        for (int col = 0; col < matrixB[0].length; col++) {
                            for (int idx = 0; idx < matrixB.length; idx++) {
                                result[row][col] = result[row][col] == null ?
                                        matrixA[row][idx] * matrixB[idx][col]
                                        : result[row][col] + matrixA[row][idx] * matrixB[idx][col];
                            }
                        }
                        latch.countDown();
                    }
                }
            });
        }
        try {
            latch.await();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        executor.shutdown();

        return result;
    }

    @Override
    public Double[][] addMatrices(Double[][] matrixA, Double[][] matrixB) {
        for (int row = 0; row < matrixA.length; row++) {
            for (int col = 0; col < matrixA[0].length; col++) {
                matrixA[row][col] = matrixA[row][col] + matrixB[row][col];
            }
        }
        return matrixA;
    }

    @Override
    public Double[][] multiplyByScalar(Double[][] matrix, Double scalar) {
        for (int row = 0; row < matrix.length; row++) {
            for (int col = 0; col < matrix[0].length; col++) {
                matrix[row][col] = matrix[row][col] * scalar;
            }
        }
        return matrix;
    }
}

File: ITMO.Python/PythonProject/Demo02/benchmark/csharp/csharp/Engine/DoubleMathematics.cs
namespace csharp.Engine
{
    internal class DoubleMathematics: Mathematics<double>
    {
        public DoubleMathematics() { }

        public override double[,] MultiplyMatricesParallel(double[,] matrixA, double[,] matrixB, int threadCount)
        {
            int rowA = matrixA.GetLength(0);
            int rowB = matrixB.GetLength(0);
            int colB = matrixB.GetLength(1);

            double[,] result = new double[rowA, colB];
            Parallel.For(0, rowA - 1, new ParallelOptions { MaxDegreeOfParallelism = threadCount }, row => {
                for (int col = 0; col < colB; col++)
                {
                    for (int idx = 0; idx < rowB; idx++)
                    {
                        result[row, col] += matrixA[row, idx] * matrixB[idx, col];
                    }
                }
            });
            return result;
        }

        public override double[,] MultiplyMatrices(double[,] matrixA, double[,] matrixB)
        {
            int rowA = matrixA.GetLength(0);
            int rowB = matrixB.GetLength(0);
            int colB = matrixB.GetLength(1);

            double[,] result = new double[rowA, colB];
            for (int row = 0; row < rowA; row++)
            {
                for (int col = 0; col < colB; col++)
                {
                    for (int idx = 0; idx < rowB; idx++)
                    {
                        result[row, col] += matrixA[row, idx] * matrixB[idx, col];
                    }
                }
            }
            return result;
        }

        public override double[,] AddMatrices(double[,] matrixA, double[,] matrixB)
        {
            int rowA = matrixA.GetLength(0);
            int colA = matrixA.GetLength(1);

            for (int row = 0; row < rowA; row++)
            {
                for (int col = 0; col < colA; col++)
                {
                    matrixA[row, col] += matrixB[row, col];
                }
            }
            return matrixA;
        }

        public override double[,] MultiplyByScalar(double[,] matrix, double scalar)
        {
            int rowCount = matrix.GetLength(0);
            int colCount = matrix.GetLength(1);

            for (int row = 0; row < rowCount; row++)
            {
                for (int col = 0; col < colCount; col++)
                {
                    matrix[row, col] *= scalar;
                }
            }
            return matrix;
        }
    }
}
